src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -23,22 +27,53 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
     /// </summary>
     [ExportRoslynLanguagesLspRequestHandlerProvider(typeof(SemanticTokensRangeHandler)), Shared]
     [Method(Methods.TextDocumentSemanticTokensRangeName)]
-    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>
+    internal class SemanticTokensRangeHandler : AbstractStatelessRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>, IDisposable
     {
         private readonly IGlobalOptionService _globalOptions;
-
-        // Lock to guard _projectToCompilation dictionary
-        private readonly object _lock = new();
-        private readonly Dictionary<ProjectId, Task<Compilation?>> _projectIdToCompilation = new();
+        private readonly IAsynchronousOperationListener _asyncListener;
 
         public override bool MutatesSolutionState => false;
         public override bool RequiresLSPSolution => true;
 
+        /// <summary>
+        /// Lock over the mutable state that follows.
+        /// </summary>
+        private readonly object _gate = new();
+
+        /// <summary>
+        /// Mapping from project id to the workqueue for producing the corresponding compilation for it on the OOP server.
+        /// </summary>
+        private readonly Dictionary<ProjectId, CompilationAvailableEventSource> _projectIdToEventSource = new();
+
+        /// <summary>
+        /// Mapping from project id to the project-cone-checksum for it we were at when the project for it had its
+        /// compilation produced on the oop server.
+        /// </summary>
+        private readonly Dictionary<ProjectId, Checksum> _projectIdToLastComputedChecksum = new();
+
+        // initialized when first request comes in.
```

Comment:
nit: maybe this would be better off as a region since it's not 100% clear just based off this comment which fields are initialized when the first request comes in

Reply:
sure!